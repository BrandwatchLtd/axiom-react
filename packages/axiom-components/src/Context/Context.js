import PropTypes from "prop-types";
import React, { Component } from "react";
import classnames from "classnames";
import Base from "../Base/Base";
import ContextTip from "./ContextTip";
import "./Context.css";

/**
 * Context provides styling elevating itself from the background.
 *
 * You normally won't need Context, consider if [Tooltip](/docs/tooltip--default) or [Dropdown](/docs/dropdown--default-story)  meet your needs first.
 * Context is a building block used within Axiom you can use it in combination with [Position](/docs/position--default) to create custom popups.
 */
export default class Context extends Component {
  static propTypes = {
    arrowRef: PropTypes.func,
    children: PropTypes.node,
    className: PropTypes.string,
    color: PropTypes.oneOf([
      "success",
      "warning",
      "error",
      "info",
      "carbon",
      "white",
    ]),
    maxHeight: PropTypes.string,
    /* passed to context tip to decide direction of arrow */
    position: PropTypes.oneOf(["top", "bottom", "left", "right"]),
    width: PropTypes.string,
  };

  static defaultProps = {
    maxHeight: "30rem",
    position: "top",
    width: "14.5rem",
  };

  render() {
    const {
      className,
      arrowRef,
      children,
      maxHeight,
      position,
      width,
      color,
      ...rest
    } = this.props;

    const classes = classnames(className, "ax-context", {
      [`ax-context--${color}`]: color,
    });
    const content = () => (
      <div className="ax-context__content" style={{ maxHeight }}>
        {children}
      </div>
    );
    return (
      <Base theme="day" {...rest} className={classes} style={{ width }}>
        {arrowRef ? (
          <ContextTip color={color} direction={position} ref={arrowRef}>
            {content()}
          </ContextTip>
        ) : (
          content()
        )}
      </Base>
    );
  }
}
