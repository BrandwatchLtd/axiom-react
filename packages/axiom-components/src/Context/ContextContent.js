import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Base from '../Base/Base';

export default class ContextContent extends Component {
  static propTypes = {
    /** Determines whether the separator below this context box should cut through the padding */
    hasFullSeparator: PropTypes.bool,
    /** Set height for the content area, box will remain at this height */
    height: PropTypes.string,
    /** Maximum height for the content area, exceeding this will make it scrollable */
    maxHeight: PropTypes.string,
    /** Padding size applied to the content area */
    padding: PropTypes.oneOf(['none', 'small', 'large']),
  };

  static defaultProps = {
    hasFullSeparator: false,
    padding: 'large',
  };

  render() {
    const { hasFullSeparator, height, maxHeight, padding, ...rest } = this.props;
    const classes = classnames(
      'ax-context-content',
      `ax-context-content--padding-${padding}`,
      {
        'ax-context-content--full-separator': hasFullSeparator,
        'ax-context-content--scrollable': maxHeight,
      }
    );

    return (
      <Base { ...rest } className={ classes } style={ { height, maxHeight } } />
    );
  }
}
