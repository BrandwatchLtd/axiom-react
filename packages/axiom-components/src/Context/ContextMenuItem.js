import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Base from '../Base/Base';
import CheckBox from '../Form/CheckBox';
import Icon from '../Icon/Icon';

export const contextMenuItemSelector = 'data-ax-context-menu-item';

export default class ContextMenuItem extends Component {
  static propTypes = {
    children: PropTypes.node,
    disabled: PropTypes.bool,
    icon: PropTypes.string,
    multiSelect: PropTypes.bool,
    onClick: PropTypes.func,
    selected: PropTypes.bool,
  };

  render() {
    const {
      children,
      disabled,
      icon,
      onClick,
      multiSelect,
      selected,
      ...rest
    } = this.props;

    const classes = classnames('ax-context-menu__item', {
      'ax-context-menu__item--selected': selected,
    });

    return (
      <Base { ...rest } { ...{ [contextMenuItemSelector]: true } }
          Component="button"
          className={ classes }
          disabled={ disabled }
          onClick={ onClick }
          textStrong={ selected }>
        { multiSelect && (
          <div className="ax-context-menu__item-checkbox">
            <CheckBox checked={ selected } disabled={ disabled } onChange={ onClick } />
          </div>
        ) }

        <div className="ax-context-menu__item-content">
          { children }
        </div>

        { !multiSelect && selected !== undefined && (
          <div className="ax-context-menu__item-icon">
            <Icon cloak={ !selected } name="tick" />
          </div>
        ) }

        { icon !== undefined && (
          <div className="ax-context-menu__item-icon">
            <Icon name={ icon } />
          </div>
        ) }
      </Base>
    );
  }
}
