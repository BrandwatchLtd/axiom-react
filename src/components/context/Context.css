:root {
  /**
   * The designs of the tip is 1.5rem from corner to corner. To calculate
   * the pre-rotated square we use `Math.sqrt((1.5rem**2)/2)`
   */
  --cmp-context-tip-size: 1.0606601717798212rem;
  --cmp-context-tip-edge-to-edge: 1.5rem;
  --cmp-context-tip-space: calc(var(--spacing-grid-size) * 2);
  --cmp-context-padding-small: calc(var(--spacing-grid-size) * 4);
  --cmp-context-padding-large: calc(var(--spacing-grid-size) * 6);
}

.ax-context {
  margin: calc(var(--cmp-context-tip-edge-to-edge) * 0.5);
  border-radius: var(--component-border-radius);
  box-shadow: var(--drop-shadow-1), var(--drop-shadow-2);
}

.ax-context__content {
  position: relative;
  border-radius: var(--component-border-radius);
  background-color: var(--color-context-background);
  color: var(--color-context-text);
  overflow-y: auto;
  /* stylelint-disable */
  /* ensures rounded corners in Safari */
  -webkit-mask-image: -webkit-radial-gradient(white, black);
  /* stylelint-enable */
}

.ax-context__arrow {
  position: absolute;
  width: var(--cmp-context-tip-size);
  height: var(--cmp-context-tip-size);
  box-shadow: var(--drop-shadow-1), var(--drop-shadow-2);
  background-color: var(--color-context-background);
  pointer-events: none;
}

.ax-context--top .ax-context__arrow,
.ax-context--bottom .ax-context__arrow {
  left: calc(50% - (var(--cmp-context-tip-size) * 0.5));
}

.ax-context--right .ax-context__arrow,
.ax-context--left .ax-context__arrow {
  top: calc(50% - (var(--cmp-context-tip-size) * 0.5));
}

.ax-context--top .ax-context__arrow {
  bottom: 0;
  transform: translate3d(0, -25%, 0) rotate(45deg);
}

.ax-context--right .ax-context__arrow {
  left: 0;
  transform: translate3d(25%, 0, 0) rotate(45deg);
}

.ax-context--bottom .ax-context__arrow {
  top: 0;
  transform: translate3d(0, 25%, 0) rotate(45deg);
}

.ax-context--left .ax-context__arrow {
  right: 0;
  transform: translate3d(-25%, 0, 0) rotate(45deg);
}
