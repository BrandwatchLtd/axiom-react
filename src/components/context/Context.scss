@import '../../design-patterns/colors/vars';
@import '../../design-patterns/colors/colors-light';
@import '../../design-patterns/spacing/vars';
@import '../vars';

$context-tip-size: $spacing-grid-size * 3;

.ax-context {
  position: relative;
  filter: drop-shadow($drop-shadow);
}

.ax-context__content {
  border-radius: $component-border-radius;
  background-color: var(--color-context-background);
  color: var(--color-context-text);
  overflow: hidden;
}

.ax-context__arrow {
  position: absolute;
  width: 0;
  height: 0;
  border-width: $context-tip-size;
  border-style: solid;
  border-color: transparent;
  pointer-events: none;
}

.ax-context--top {
  margin-bottom: $context-tip-size;
}

.ax-context--bottom {
  margin-top: $context-tip-size;
}

.ax-context--right {
  margin-left: $context-tip-size;
}

.ax-context--left {
  margin-right: $context-tip-size;
}

.ax-context--top,
.ax-context--bottom {
  .ax-context__arrow {
    left: calc(50% - #{$context-tip-size});
  }
}

.ax-context--right,
.ax-context--left {
  .ax-context__arrow {
    top: calc(50% - #{$context-tip-size});
  }
}

.ax-context--top {
  .ax-context__arrow {
    bottom: -($context-tip-size * 2);
    border-top-color: var(--color-context-background);
  }
}

.ax-context--right {
  .ax-context__arrow {
    left: -($context-tip-size * 2);
    border-right-color: var(--color-context-background);
  }
}

.ax-context--bottom {
  .ax-context__arrow {
    top: -($context-tip-size * 2);
    border-bottom-color: var(--color-context-background);
  }
}

.ax-context--left {
  .ax-context__arrow {
    right: -($context-tip-size * 2);
    border-left-color: var(--color-context-background);
  }
}
