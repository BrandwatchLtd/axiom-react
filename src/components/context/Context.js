import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import { Base } from 'bw-axiom';
import './Context.css';

export default class Context extends Component {
  static propTypes = {
    arrowRef: PropTypes.func,
    children: PropTypes.node,
    position: PropTypes.oneOf(['top', 'bottom', 'right', 'left']),
    theme: PropTypes.oneOf(['dark', 'light']),
    width: PropTypes.string,
  };

  static defaultProps = {
    position: 'top',
    theme: 'light',
    width: '14.5rem',
  };

  render() {
    const { arrowRef, children, position, theme, width, ...rest } = this.props;
    const style = { width };
    const classes = classnames('ax-context',
      `ax-context--${position}`,
      `ax-context--${theme}`);

    return (
      <Base { ...rest } className={ classes } style={ style }>
        <div className="ax-context__content">
          { children }
        </div>

        <span
            className="ax-context__arrow"
            ref={ arrowRef } />
      </Base>
    );
  }
}
