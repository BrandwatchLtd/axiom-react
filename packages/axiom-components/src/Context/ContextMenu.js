import PropTypes from 'prop-types';
import React, { Component } from 'react';
import ContextContent from './ContextContent';

export default class ContextMenu extends Component {
  static propTypes = {
    children: PropTypes.node,
  };

  render() {
    const { children, ...rest } = this.props;

    return (
      <ContextContent { ...rest } padding="none">
        <div className="ax-context-menu">
          { children }
        </div>
      </ContextContent>
    );
  }
}
