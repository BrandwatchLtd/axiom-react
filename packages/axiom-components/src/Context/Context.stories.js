import React from "react";
import Context from "./Context";
import ContextContent from "./ContextContent";
import ContextMenu from "./ContextMenu";
import ContextMenuItem from "./ContextMenuItem";

export default {
  title: "Context",
  component: Context,
  subcomponents: { ContextContent, ContextMenu, ContextMenuItem },
};

export function Default(props) {
  return <Context {...props}>Default Context</Context>;
}

export function WithContextContent(props) {
  return (
    <Context {...props}>
      <ContextContent paddingVertical="medium" height="50px">
        Good Job
      </ContextContent>
    </Context>
  );
}

WithContextContent.args = { color: "success" };

WithContextContent.parameters = {
  docs: {
    description: {
      story: "ContextContent allows you to configure the contents appearance.",
    },
  },
};

export function WithContextMenu({ ...rest }) {
  return (
    <Context {...rest}>
      <ContextMenu paddingVertical="medium">
        <ContextMenuItem>Top</ContextMenuItem>
        <ContextMenuItem selected>Middle</ContextMenuItem>
        <ContextMenuItem>Bottom</ContextMenuItem>
      </ContextMenu>
    </Context>
  );
}

WithContextMenu.parameters = {
  docs: {
    description: {
      story: "ContextMenu and ContextMenuItem can be used to help build menus.",
    },
  },
};
