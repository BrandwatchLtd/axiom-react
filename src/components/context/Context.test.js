import React from 'react';
import renderer from 'react-test-renderer';
import Context from './Context';

const getComponent = (props = {}) =>
  renderer.create(
    <Context { ...props }>
      Lorem ipsum
    </Context>
  );

describe('Context', () => {
  it('renders with defaultProps', () => {
    const component = getComponent();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with arrowRef', () => {
    const component = getComponent({ arrowRef: () => {} });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders custom maxHeight', () => {
    const component = getComponent({ maxHeight: '20rem' });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders custom width', () => {
    const component = getComponent({ width: '20rem' });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('renders with position', () => {
    ['top', 'bottom', 'right', 'left'].forEach((position) => {
      it(position, () => {
        const component = getComponent({ position });
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe('renders with theme', () => {
    ['dark', 'light'].forEach((theme) => {
      it(theme, () => {
        const component = getComponent({ theme });
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
