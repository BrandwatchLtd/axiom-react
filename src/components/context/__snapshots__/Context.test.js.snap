// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context renders custom maxHeight 1`] = `
<div
  className="ax-context ax-context--top ax-theme--light"
  style={
    Object {
      "width": "14.5rem",
    }
  }
>
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "20rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;

exports[`Context renders custom width 1`] = `
<div
  className="ax-context ax-context--top ax-theme--light"
  style={
    Object {
      "width": "20rem",
    }
  }
>
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "30rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;

exports[`Context renders with arrowRef 1`] = `
<div
  className="ax-context ax-context--top ax-theme--light"
  style={
    Object {
      "width": "14.5rem",
    }
  }
>
  <span
    className="ax-context__arrow"
  />
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "30rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;

exports[`Context renders with defaultProps 1`] = `
<div
  className="ax-context ax-context--top ax-theme--light"
  style={
    Object {
      "width": "14.5rem",
    }
  }
>
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "30rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;

exports[`Context renders with position bottom 1`] = `
<div
  className="ax-context ax-context--bottom ax-theme--light"
  style={
    Object {
      "width": "14.5rem",
    }
  }
>
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "30rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;

exports[`Context renders with position left 1`] = `
<div
  className="ax-context ax-context--left ax-theme--light"
  style={
    Object {
      "width": "14.5rem",
    }
  }
>
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "30rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;

exports[`Context renders with position right 1`] = `
<div
  className="ax-context ax-context--right ax-theme--light"
  style={
    Object {
      "width": "14.5rem",
    }
  }
>
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "30rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;

exports[`Context renders with position top 1`] = `
<div
  className="ax-context ax-context--top ax-theme--light"
  style={
    Object {
      "width": "14.5rem",
    }
  }
>
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "30rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;

exports[`Context renders with theme dark 1`] = `
<div
  className="ax-context ax-context--top ax-theme--dark"
  style={
    Object {
      "width": "14.5rem",
    }
  }
>
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "30rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;

exports[`Context renders with theme light 1`] = `
<div
  className="ax-context ax-context--top ax-theme--light"
  style={
    Object {
      "width": "14.5rem",
    }
  }
>
  <div
    className="ax-context__content"
    style={
      Object {
        "maxHeight": "30rem",
      }
    }
  >
    Lorem ipsum
  </div>
</div>
`;
